x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java

```diff
@@ -184,6 +187,32 @@ synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task,
             failure, isPartial, frozen == false, task.getStartTime(), expirationTime);
     }
 
+
+    /**
+     * Creates an {@link AsyncStatusResponse} -- status of an async response.
+     * Response is created based on the current state of the mutable response or based on {@code finalResponse} if it is available.
+     * @param asyncExecutionId – id of async search request
+     * @param startTime – start time of task
+     * @param expirationTime – expiration time of async search request
+     * @return response representing the status of async search
+     */
+    AsyncStatusResponse toStatusResponse(String asyncExecutionId, long startTime, long expirationTime) {
+        if (frozen == false) {
+            return new AsyncStatusResponse(
+                asyncExecutionId,
+                startTime,
+                expirationTime,
+                finalResponse != null ? finalResponse.getTotalShards() : totalShards,
+                finalResponse != null ? finalResponse.getSuccessfulShards() : successfulShards,
+                finalResponse != null ? finalResponse.getSkippedShards() : skippedShards,
+                finalResponse != null ? (finalResponse.getShardFailures() == null ? finalResponse.getShardFailures().length : 0) :
+                    (queryFailures != null ? queryFailures.length() : 0)
+            );
+        } else {
+            return getCompletedSearchStatusResponse(asyncExecutionId, expirationTime);
```

Comment:
We still need the number of shards, start time, ... I don't understand why the status would be reseted to 0 when the search is done ?

Reply:
So I think you don't need to check `frozen` and just build the status from the finalResponse if it's available and from the partial stats otherwise.

Reply:
addressed in 0775cf1 with the new format for a completed status.